velox/functions/sparksql/Rand.h

```diff
@@ -23,54 +23,25 @@ template <typename T>
 struct RandFunction {
   static constexpr bool is_deterministic = false;
 
-  FOLLY_ALWAYS_INLINE void call(double& result) {
-    result = folly::Random::randDouble01();
+  template <typename TInput>
+  void initialize(const core::QueryConfig& config, const TInput* seedInput) {
+    auto partitionId = config.sparkPartitionId();
+    generator_ = std::mt19937{};
+    int64_t seed = seedInput ? (int64_t)*seedInput : 0;
+    generator_.seed(seed + partitionId);
   }
 
-  FOLLY_ALWAYS_INLINE void callNullable(
-      double& result,
-      const int32_t* seed,
-      const int32_t* partitionIndex) {
-    initializeGenerator(seed, partitionIndex);
-    result = folly::Random::randDouble01(*generator_);
-  }
-
-  // To differentiate generator for each thread, seed plus partitionIndex is
-  // the actual seed used for generator.
-  FOLLY_ALWAYS_INLINE void callNullable(
-      double& result,
-      const int64_t* seed,
-      const int32_t* partitionIndex) {
-    initializeGenerator(seed, partitionIndex);
-    result = folly::Random::randDouble01(*generator_);
+  FOLLY_ALWAYS_INLINE void call(double& result) {
+    result = folly::Random::randDouble01();
   }
 
-  // For NULL constant input of unknown type.
-  FOLLY_ALWAYS_INLINE void callNullable(
-      double& result,
-      const UnknownValue* /*seed*/,
-      const int32_t* partitionIndex) {
-    initializeGenerator<int64_t>(nullptr, partitionIndex);
-    result = folly::Random::randDouble01(*generator_);
+  template <typename TInput>
+  FOLLY_ALWAYS_INLINE void callNullable(double& result, TInput /*seedInput*/) {
```

Comment:
Why do we ignore the 'seedInput' parameter? Is it supposed to be constant?

Reply:
Yes, the rand seed should be constant. https://github.com/apache/spark/blob/3aa16e8193cf59379c4ad78c5f0a7b6490aec4b8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala#L396